odoo/models.py

```diff
@@ -1731,19 +1768,326 @@ def clear_caches(cls):
+        :raise AccessError: if user is not allowed to access requested information """
+
+        self.check_access_rights('read')
+
+        groupby = [groupby] if isinstance(groupby, str) else groupby
+        aggregates = [aggregates] if isinstance(aggregates, str) else aggregates
+
+        if expression.is_false(self, domain):
+            if not groupby:  # If there is no group postgresql always return a row
+                return [tuple(self._read_group_empty_value(spec) for spec in itertools.chain(groupby, aggregates))]
+            return []
+
+        order_traverse_many2one = True
+        if not order:
+            order = ','.join(groupby)
+            order_traverse_many2one = False
+
+        query = self._search(domain)
+
+        fnames_to_flush = OrderedSet()
+        groupby_terms = []  # [<SQL expression>,]
+        for spec in groupby:
+            sql_expression, fnames_used = self._read_group_groupby(query, spec)
+            fnames_to_flush.update(fnames_used)
+            groupby_terms.append(sql_expression)
+
+        select_terms = []  # [<SQL expression>,]
+        for spec in aggregates:
+            sql_expression, fnames_used = self._read_group_select(query, spec)
+            fnames_to_flush.update(fnames_used)
+            select_terms.append(sql_expression)
+
+        having_expression, having_params, fnames_used = self._read_group_having(query, having)
+        fnames_to_flush.update(fnames_used)
+
+        orderby_terms, extra_groupby_terms, fnames_used = self._read_group_orderby(query, groupby, order, order_traverse_many2one)
+        fnames_to_flush.update(fnames_used)
+
+        from_clause, where_clause, query_params = query.get_sql()
+        query_parts = [
+            f"SELECT {', '.join(groupby_terms + select_terms)}",
+            f"FROM {from_clause}",
+        ]
+        if where_clause:
+            query_parts.append(f"WHERE {where_clause}")
+        if groupby_terms:
+            query_parts.append(f"GROUP BY {', '.join(groupby_terms + extra_groupby_terms)}")
+        if having_expression:
+            query_parts.append(f"HAVING {having_expression}")
+            query_params.extend(having_params)
+        if orderby_terms:
+            query_parts.append(f"ORDER BY {', '.join(orderby_terms)}")
+        if limit:
+            query_parts.append("LIMIT %s")
+            query_params.append(limit)
+        if offset:
+            query_parts.append("OFFSET %s")
+            query_params.append(offset)
+
+        self._flush_search(domain, fnames_to_flush)
+        if fnames_to_flush:
+            self._read_group_check_field_access_rights(fnames_to_flush)
+
+        self.env.cr.execute('\n' + '\n'.join(query_parts), query_params)
+
+        return list(zip(*(
+            self._read_group_postprocess(spec, values)
+            for spec, values in zip(itertools.chain(groupby, aggregates), zip(*self.env.cr.fetchall()))
+        )))
```

Comment:
> The double transposition and 3-deep zip is pretty rough.

Yes, you are right. It is too complicate to read. I will try to find something else.

> So if I understand correctly this basically converts the output of GROUP_BY into a column-major form, post-processes each column according to its spec, then transposes back to row-major for the final output?

Not only the `GROUP_BY` but also the aggregates selected.

> Since _read_group_postprocess just maps over the values anyway, wouldn't it be easier (and more readable) to create a sequence of post-processing functions then apply those to the group rows? Unless there are plants to leverage the column-major traversal but that seems iffy.

Yep, I will try to have a more readable code.